toolchain/semantics/semantics_context.h

```diff
@@ -161,6 +171,38 @@ class SemanticsContext {
     Compatible,
   };
 
+  // Provides DenseMap support for canonical_struct_types_.
+  struct StructTypeMapInfo {
+    using KeyType =
+        llvm::SmallVector<std::pair<SemanticsStringId, SemanticsTypeId>>;
+    using IntMapInfo = llvm::DenseMapInfo<int32_t>;
+
+    static inline auto getEmptyKey() -> KeyType {
+      return {{SemanticsStringId(IntMapInfo::getEmptyKey()),
+               SemanticsTypeId(IntMapInfo::getEmptyKey())}};
+    }
+
+    static inline auto getTombstoneKey() -> KeyType {
+      return {{SemanticsStringId(IntMapInfo::getTombstoneKey()),
+               SemanticsTypeId(IntMapInfo::getTombstoneKey())}};
+    }
+
+    static auto getHashValue(const KeyType& val) -> unsigned {
+      unsigned hash_value = 0;
+      for (const auto& [string_id, type_id] : val) {
+        hash_value = llvm::detail::combineHashValue(
+            hash_value, IntMapInfo::getHashValue(string_id.index));
+        hash_value = llvm::detail::combineHashValue(
+            hash_value, IntMapInfo::getHashValue(type_id.index));
+      }
+      return hash_value;
+    }
```

Comment:
All of this may prove really inefficient, as in go quadratic or other sharp scaling cliffs that cause problems long before we get to benchmarking... DenseMap assumes its keys are reasonably small and easy to build.

LLVM actually has a utility for building exactly this kind of data structure that is optimized for the case you're handling: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-foldingset-h

However, that type is *also* optimized for the case of a polymorphic object as the nodes, which you don't need. Not sure which will end up being dominant here.

Reply:
Switched to FoldingSet.